x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.ts

```diff
@@ -353,12 +386,25 @@ export const performBulkActionRoute = (
         let created: RuleAlertType[] = [];
         let deleted: RuleAlertType[] = [];
 
+        const throwDryRunMlAuth = (ruleType: Type) =>
+          throwDryRunError(
+            async () => throwAuthzError(await mlAuthz.validateRuleType(ruleType)),
+            BULK_ACTIONS_DRY_RUN_ERR_CODE.MACHINE_LEARNING_AUTH
+          );
+
         switch (body.action) {
           case BulkAction.enable:
             bulkActionOutcome = await initPromisePool({
               concurrency: MAX_RULES_TO_UPDATE_IN_PARALLEL,
               items: rules,
               executor: async (rule) => {
+                if (isDryRun) {
+                  if (!rule.enabled) {
+                    await throwDryRunMlAuth(rule.params.type);
+                  }
+                  return rule;
+                }
```

Comment:
> It violates the separation of concerns because the route handler on the top level becomes aware of the implementation details of the methods it calls.

Can you please elaborate on this? Route handler takes different payload and process it depends on payload: see bulk actions switch. New query parameter it's just another value for payload

> As with any duplication, we should keep in sync request validation logic for both code paths: dry run and normal execution. But, even if we sync it right now, how will we ensure that it wouldn't diverge in the future?
How do we ensure that we've covered all situations that result in an error? In this PR, three error types are handled: immutable rules, ML, and auth. Is that list exhaustive?

That's a trade-off I had to make here.
Let's check bulk edit action: it calls `rulesClient.bulkEdit`:

```ts
        // handling this action before switch statement as bulkEditRules fetch rules within
        // rulesClient method, hence there is no need to use fetchRulesByQueryOrIds utility
        if (body.action === BulkAction.edit && !isDryRun) {
          const { rules, errors } = await bulkEditRules({
            rulesClient,
            filter: query,
            ids: body.ids,
            actions: body.edit,
            mlAuthz,
          });

```

How do we know what's there and what kind of validation done in that method? We can't go inside it's implementation and track it.
So, instead of I've approached this from a business view standpoint: when we need to throw error?
In our case: 
- for ML rules when index patterns action applied
- for immutable rules
- auth

So, I explicitly covered this in dry_run implementation.

- How to ensure it's up to date and synced?
 **Only with test in my opinion.**
- Is that list of errors exhaustive? 
 **It's not, but for now covers all our needs. If some error will be thrown during actual run, user still be able to see it. But main cases will be covered. And we extend them in future**
- Why error codes? 
 **To make sure errors classified and can be easily processed**
- Should error codes be added in regular run? 
 **Probably, ideally. But because we approach it from business view standpoint and we can't dig deep into internal API implementation, we should be fine to make this trade-off**

Reply:
as agreed, validation logic moved to a validation function, that called once during processing bulk action